cilium/secret_watcher.cc

```diff
@@ -0,0 +1,179 @@
+#include "cilium/secret_watcher.h"
+
+#include "source/common/config/datasource.h"
+
+#include "cilium/grpc_subscription.h"
+
+namespace Envoy {
+namespace Cilium {
+
+namespace {
+
+// SDS config used in production
+envoy::config::core::v3::ConfigSource getCiliumSDSConfig(const std::string&) {
+  /* returned config_source has initial_fetch_timeout left at default 15 seconds. */
+  return Cilium::cilium_xds_api_config;
+}
+
+Secret::GenericSecretConfigProviderSharedPtr
+secretProvider(Server::Configuration::TransportSocketFactoryContext& context,
+               const std::string& sds_name) {
+  envoy::config::core::v3::ConfigSource config_source = getSDSConfig(sds_name);
+  return context.secretManager().findOrCreateGenericSecretProvider(config_source, sds_name, context,
+                                                                   context.initManager());
+}
+
+} // namespace
+
+getSDSConfigFunc getSDSConfig = &getCiliumSDSConfig;
+void setSDSConfigFunc(getSDSConfigFunc func) { getSDSConfig = func; }
+
+SecretWatcher::SecretWatcher(const NetworkPolicyMap& parent, const std::string& sds_name)
+    : parent_(parent), name_(sds_name),
+      secret_provider_(secretProvider(parent.transportFactoryContext(), sds_name)),
+      update_secret_(readAndWatchSecret()) {}
+
+SecretWatcher::~SecretWatcher() { delete load(); }
+
+Envoy::Common::CallbackHandlePtr SecretWatcher::readAndWatchSecret() {
+  store();
+  return secret_provider_->addUpdateCallback([this]() { store(); });
+}
+
+void SecretWatcher::store() {
+  const auto* secret = secret_provider_->secret();
+  if (secret != nullptr) {
+    Api::Api& api = parent_.transportFactoryContext().api();
+    std::string* p = new std::string(Config::DataSource::read(secret->secret(), true, api));
+    ENVOY_LOG(debug, "allocated new secret value: {} ({})", static_cast<void*>(p), *p);
```

Comment:
any change the secret value is leaked in debug mode here ?

Reply:
Yes, thanks for pointing this out. This is a leftover from local testing that should not be needed anymore.